packages/valory/skills/staking_abci/behaviours.py

```diff
@@ -0,0 +1,323 @@
+        """Check whether the service is staked."""
+        if self.synchronized_data.period_count != 0:
+            self.is_service_staked = self.synchronized_data.is_service_staked
+            return True
+
+        service_id = self.params.on_chain_service_id
+        if service_id is None:
+            self.context.logger.warning(
+                "Cannot perform any staking-related operations without a configured on-chain service id. "
+                "Setting status to 'not staked'."
+            )
+            return True
+
+        status = yield from self._staking_contract_interact(
+            contract_callable="is_service_staked",
+            placeholder=get_name(CallCheckpointBehaviour.is_service_staked),
+            service_id=service_id,
+        )
+
+        return status
+
+    def _get_next_checkpoint(self) -> WaitableConditionType:
+        """Get the timestamp in which the next checkpoint is reached."""
+        status = yield from self._staking_contract_interact(
+            contract_callable="get_next_checkpoint_ts",
+            placeholder=get_name(CallCheckpointBehaviour.next_checkpoint),
+        )
+        return status
+
+    def _build_checkpoint_tx(self) -> WaitableConditionType:
+        """Get the request tx data encoded."""
+        result = yield from self._staking_contract_interact(
+            contract_callable="build_checkpoint_tx",
+            placeholder=get_name(CallCheckpointBehaviour.checkpoint_data),
+        )
+
+        return result
+
+    def _get_safe_tx_hash(self) -> WaitableConditionType:
+        """Prepares and returns the safe tx hash."""
+        status = yield from self.contract_interact(
+            contract_address=self.synchronized_data.safe_contract_address,
+            contract_public_id=GnosisSafeContract.contract_id,
+            contract_callable="get_raw_safe_transaction_hash",
+            data_key="tx_hash",
+            placeholder=get_name(CallCheckpointBehaviour.safe_tx_hash),
+            to_address=self.params.staking_contract_address,
+            value=ETH_PRICE,
+            data=self.checkpoint_data,
+        )
+        return status
+
+    def _prepare_safe_tx(self) -> Generator[None, None, str]:
+        """Prepare the safe transaction for calling the checkpoint and return the hex for the tx settlement skill."""
+        yield from self.wait_for_condition_with_sleep(self._build_checkpoint_tx)
+        yield from self.wait_for_condition_with_sleep(self._get_safe_tx_hash)
+        return hash_payload_to_hex(
+            self.safe_tx_hash,
+            ETH_PRICE,
+            SAFE_GAS,
+            self.params.staking_contract_address,
+            self.checkpoint_data,
+        )
+
+    def async_act(self) -> Generator:
+        """Do the action."""
+        with self.context.benchmark_tool.measure(self.behaviour_id).local():
+            yield from self.wait_for_condition_with_sleep(self._check_service_staked)
```

Comment:
This doesn't make sense for non-staking services.

I would have preferred non-staking services never reach this FSM. But given this design, you need at least some sort of boolean to track if this service even needs to do anything here.

Reply:
I mean now that it is implemented this way, it will work..

Reply:
It does have a boolean in the synced data and checks on-chain only during the first period:

https://github.com/valory-xyz/trader/blob/908bc84b6f9832d795d6700a4115c1cda402b4fa/packages/valory/skills/staking_abci/behaviours.py#L230-L232

If not staked, it does nothing.

As mentioned above, it would not make much sense conceptually to check whether the service is staked in any of the other abcis.
Moreover, this implementation makes this component as much independent as possible, so that future services can easily make use of it.